File: DSA/src/go/find_the_duplicate_number.go
package main

func findDuplicate(nums []int) int {
	var slow, fast int
	for fast < len(nums) {
		slow = nums[slow]
		fast = nums[nums[fast]]
		if slow == fast {
			break
		}
	}

	first, second := 0, slow
	for first != second {
		first = nums[first]
		second = nums[second]
	}
	return first
}

File: docs/DSA/src/py/find_the_duplicate_number.py
class Solution:
    def findDuplicate(self, nums: list[int]) -> int:
        slow, fast = 0, 0
        while fast < len(nums):
            slow = nums[slow]
            fast = nums[nums[fast]]
            if slow == fast:
                break

        first, second = 0, slow
        while first != second:
            first = nums[first]
            second = nums[second]
        return first